### PR TITLE
Price calculator skeleton for JS

### PR DESCRIPTION
**Pull request by cosmini**
## Motivation

We have lots of Fullstack candidates that don't know any of the already provided languages but are proficient in JS.
We should provide the skeleton in JS so they can have a fair go at this exercise.
This is a transliteration from python skeleton (and some help from go skeleton).

## Example

```
node bin/price-calculator --cart ../../testdata/cart-4560.json --base-prices ../../testdata/base-prices.json
4560
```

### CODE REVIEW

**Review by tklarryonline — COMMENTED** [review]
Is this a skeleton version (like Python) or a fully functional Price Calculator app? Also, should we move all the files into `js/price_calculator` for consistency?

File: fullstack/js/cli.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import arg from 'arg'
+import inquirer from 'inquirer'
+import calculate from './src/calculate'
+import chalk from 'chalk'
+import { round } from 'lodash'
+
+/**
+ * Get card and price from command line arguments.
```
Comment (tsommer) [inline fullstack/js/cli.js]: `card` -> `cart`

**Review by tsommer — COMMENTED** [review]
Let's also add a change to the main `fullstack` for this one, to mention `Javascript` as one of the coding options.

And agree with Luan about using the same folder structure as the other languages

**Comment by tsommer** [comment]
@tklarryonline - it looks like this follows the same pattern as `Ruby` and `Go`... which in my books is OK for the moment to make sure we get this out the door

We still need to sit together and decide what general approach we'd like to take, but let's do that once we have a JS option available

**Comment by tklarryonline** [comment]
@tsommer I was confused by the PR name. I also agree that taking either approach for the JS exercise is Okay for me.

@cosmini Let's schedule some time for us to take a look at how to do interviews with the JS option.

**Comment by cosmini** [comment]
> fully functional P

**Comment by cosmini** [comment]
Ignore my fat finger actions 😅

**Comment by cosmini** [comment]
> Is this a skeleton version (like Python) or a fully functional Price Calculator app? Also, should we move all the files into `js/price_calculator` for consistency?
@tklarryonline 
it is fully functional price calculator (but a good skeleton for the extension exercise) so I got a bit confused by the skeleton term.
I've updated the title.

As for the discussion, I've sent an invite for us and Blake.

**Comment by cosmini** [comment]
@tsommer @tklarryonline This have been updated to follow the python skeleton as much as possible. Can you please re-review it ?

File: fullstack/js/price-calculator/app/managers/__tests__/base-price-manager.spec.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  })
+
+  it('should lookup zero option item', () => {
+    expect.assertions(1)
+
+    const result = basePriceManager.lookup('leggings')
+
+    expect(result).toStrictEqual(
+      new ProductBasePrice({
+        'product-type': 'hoodie',
```
Comment (tklarryonline) [inline fullstack/js/price-calculator/app/managers/__tests__/base-price-manager.spec.js]: Should this be `leggings`?

```suggestion
        'product-type': 'leggings',
```

File: fullstack/js/price-calculator/app/managers/__tests__/base-price-manager.spec.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        options: {},
+        'base-price': 5000,
+      })
+    )
```
Comment (tklarryonline) [inline fullstack/js/price-calculator/app/managers/__tests__/base-price-manager.spec.js:60]: I think it might be beneficial to add a test case for zero option hoodie that would return nothing.

File: fullstack/js/price-calculator/app/managers/base-price-manager.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+  constructor(jsonFile) {
+    this.data = this.readJson(jsonFile)
+  }
+
+  readJson(jsonFile) {
+    let data = {}
+    const rawData = require(jsonFile)
+
+    for (const rawItem in rawData) {
```
Comment (tklarryonline) [inline fullstack/js/price-calculator/app/managers/base-price-manager.js:14]: Should this be `of`?

File: fullstack/js/price-calculator/app/managers/cart-manager.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    this.data = this.readJson(jsonFile)
+    this.basePriceManager = basePriceManager
+  }
+
+  readJson(jsonFile) {
+    let data = []
+    const rawData = require(jsonFile)
+
+    for (const rawItem in rawData) {
+      const cartItem = new CartItem(rawItem)
```
Comment (tklarryonline) [inline fullstack/js/price-calculator/app/managers/cart-manager.js]: Should the for-loop use `of`? Also, the `CartItem` constructor wouldn't work here I think

File: fullstack/js/price-calculator/app/managers/base-price-manager.js (hunk 4 above)
Comment (cosmini) [reply fullstack/js/price-calculator/app/managers/base-price-manager.js:14]: According to MDN `of` can iterate Array, Object, Map, Set, arguments while `in` only arrays.
Since `rawData` is array we can leave this as it is ?
Source: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Loops_and_iteration#for...of_statement

File: fullstack/js/price-calculator/app/managers/cart-manager.js (hunk 5 above)
Comment (cosmini) [reply fullstack/js/price-calculator/app/managers/cart-manager.js]: Yes, I'll have to provide arguments to `CartItem` 👍

File: fullstack/js/price-calculator/app/managers/__tests__/base-price-manager.spec.js (hunk 3 above)
Comment (cosmini) [reply fullstack/js/price-calculator/app/managers/__tests__/base-price-manager.spec.js:60]: Do we need to add a test case like this to the other skeletons?

Comment (tklarryonline) [reply fullstack/js/price-calculator/app/managers/__tests__/base-price-manager.spec.js:60]: Let's leave it. We can add later if we want

File: fullstack/js/price-calculator/app/managers/base-price-manager.js (hunk 4 above)
Comment (tklarryonline) [reply fullstack/js/price-calculator/app/managers/base-price-manager.js:14]: Somehow I thought that `rawData` is an object. Good to me then

**Review by tklarryonline — APPROVED** [review]
